sup3r/pipeline/forward_pass.py

```diff
@@ -1738,61 +1773,81 @@ def run(cls, strategy, node_index):
             logger.debug(f'Running forward passes on node {node_index} in '
                          'serial.')
             for i, chunk_index in enumerate(strategy.node_chunks[node_index]):
                 now = dt.now()
-                cls.incremental_check_run(strategy, node_index, chunk_index)
+                fwp = cls.incremental_check_load(strategy=strategy,
+                                                 node_index=node_index,
+                                                 chunk_index=chunk_index)
+                cls._run_fwp_or_future(fwp)
                 mem = psutil.virtual_memory()
                 logger.info('Finished forward pass on chunk_index='
                             f'{chunk_index} in {dt.now() - now}. {i + 1} of '
                             f'{len(strategy.node_chunks[node_index])} '
                             'complete. Current memory usage is '
                             f'{mem.used / 1e9:.3f} GB out of '
                             f'{mem.total / 1e9:.3f} GB total.')
+            logger.info('Finished forward passes on '
+                        f'{len(strategy.node_chunks[node_index])} chunks in '
+                        f'{dt.now() - start}')
 
         else:
-            logger.debug(f'Running forward passes on node {node_index} in '
-                         'parallel with pass_workers='
-                         f'{strategy.pass_workers}.')
-            with SpawnProcessPool(max_workers=strategy.pass_workers) as exe:
-                futures = {}
+            logger.info('Running asynchronous forward passes on node '
+                        f'{node_index} in parallel with pass_workers='
+                        f'{strategy.pass_workers}.')
+
+            load_futures = {}
+            fwp_futures = {}
+            with ThreadPoolExecutor(max_workers=strategy.pass_workers) as exe:
                 now = dt.now()
-                for chunk_index in strategy.node_chunks[node_index]:
-                    future = exe.submit(cls.incremental_check_run,
-                                        strategy=strategy,
-                                        node_index=node_index,
-                                        chunk_index=chunk_index)
-                    futures[future] = chunk_index
-
-                logger.info(f'Started {len(futures)} forward passes '
+                for i, chunk_index in enumerate(
+                        strategy.node_chunks[node_index]):
+                    load_future = exe.submit(cls.incremental_check_load,
+                                             strategy=strategy,
+                                             node_index=node_index,
+                                             chunk_index=chunk_index)
+                    load_futures[load_future] = chunk_index
+                    fwp_future = exe.submit(cls._run_fwp_or_future,
+                                            load_future)
+                    fwp_futures[fwp_future] = chunk_index
```

Comment:
This is an interesting implementation, my only concern is how do you ensure that the 2 workers are working on a load future and a fwp future in parallel? This isn't enforced is it? I would think you could get scenarios where you get the 2 workers doing only 2x loads or only 2x fwps, thereby not really balancing the IO with the compute and possibly running into memory constraints.

Reply:
Yeah my initial implementation ran exe.submit on fwp instantiations and returned those futures. Another thread went through those futures and when one was completed ran the corresponding fwp.run_chunk(). This was a more explicit separation but was not nearly the speed up we see here. I've been using this to run all the wtk/seasia runs and havent seen a mem error yet

Reply:
Gotcha, so current implementation actually does an implicit dependency graph so that when pass_workers=2 it will only work on a single load+fwp at any given time because concurrent futures can only march through the compute queue in order. Looks good!